lib/modules/manager/sbt/extract.ts

```diff
@@ -323,23 +373,123 @@ export function extractPackageFile(
 
   try {
     parsedResult = scala.query(content, query, {
-      vars: {},
+      globalVars,
+      localVars,
       deps: [],
-      registryUrls: [REGISTRY_URLS.mavenCentral],
+      registryUrls,
+      packageFile,
+      scalaVersion,
     });
   } catch (err) /* istanbul ignore next */ {
-    logger.debug({ err, packageFile }, 'Sbt parsing error');
+    logger.warn({ err, packageFile }, 'Sbt parsing error');
   }
 
   if (!parsedResult) {
     return null;
   }
 
-  const { deps, packageFileVersion } = parsedResult;
+  delete parsedResult?.globalVars;
+  return parsedResult;
+}
 
-  if (!deps.length) {
-    return null;
+function prepareLoadPackageFiles(
+  packageFilesContent: { packageFile: string; content: string }[]
+): Pick<Ctx, 'globalVars' | 'registryUrls' | 'scalaVersion'> {
+  // Return variable
+  let globalVars: Ctx['globalVars'] = {};
+  let registryUrls: Ctx['registryUrls'] = [REGISTRY_URLS.mavenCentral];
+  let scalaVersion: Ctx['scalaVersion'] = undefined;
+
+  for (const { packageFile, content } of packageFilesContent) {
+    const res = extractDependency(content, {
+      deps: [],
+      registryUrls,
+      localVars: globalVars,
+      packageFile,
+    });
+
+    if (res) {
+      globalVars = { ...globalVars, ...res.localVars };
+      if (res.registryUrls) {
+        registryUrls = [...registryUrls, ...res.registryUrls];
+      }
+      if (res.scalaVersion) {
+        scalaVersion = res.scalaVersion;
+      }
+    }
+  }
+
+  return {
+    globalVars,
+    registryUrls,
+    scalaVersion,
+  };
+}
+
+export async function extractAllPackageFiles(
+  _config: ExtractConfig,
+  packageFiles: string[]
+): Promise<PackageFile[] | null> {
+  // Read packages and store in groupPackageFileContent
+  // group package file by its folder
+  const groupPackageFileContent: GroupFilenameContent = {};
+  for (const packageFile of packageFiles) {
+    const content = await readLocalFile(packageFile, 'utf8');
+    if (!content) {
+      logger.trace({ packageFile }, 'packageFile has no content');
+      continue;
+    }
+    const group = upath.dirname(packageFile);
+    groupPackageFileContent[group] ??= [];
+    groupPackageFileContent[group].push({ packageFile, content });
+  }
+
+  // 1. globalVariables from project/ and root package file
+  // 2. registry from all package file
+  // 3. Project's scalaVersion - to add suffix eg. "_2.13"
+  const { globalVars, registryUrls, scalaVersion } = prepareLoadPackageFiles([
+    ...(groupPackageFileContent['project'] ?? []), // in project/ folder
+    ...(groupPackageFileContent['.'] ?? []), // root
+  ]);
+
+  const mapDepsToPackageFile: Record<
+    string,
+    PackageDependency<SbtManagerData>[]
+  > = {};
+  // Start extract all package files
+  for (const packageFileContents of Object.values(groupPackageFileContent)) {
+    for (const { packageFile, content } of packageFileContents) {
+      const res = extractDependency(content, {
```

Comment:
See comment above. This is very inefficient because with `scala.query` in `extractDependency`, you also extract `globalVars`, `localVars`, `registryUrls`, again which you then throw away here. I'd rather make two separate queries, so that you don't match absolutely **everything** twice.